client/core/bond.go

```diff
@@ -39,58 +40,225 @@ func (c *Core) watchExpiredBonds(ctx context.Context) {
+// rules and block times for an asset. For some assets this could be zero, while
+// for others like Bitcoin, a time-locked output becomes spendable when the
+// median of the last 11 blocks is greater than the lockTime. This function
+// returns a high value to avoid all but extremely rare (but temporary) drops in
+// tier. NOTE: to minimize bond overlap, an asset.Bonder method could provide
+// this estimate, but it is still very short relative to the entire bond
+// lifetime, which is on the order of months.
+func spendableDelay(net dex.Network) int64 {
+	// We use 3*pendingBuffer as a well-padded estimate of this duration. e.g.
+	// with mainnet, we would use a 90 minute pendingBuffer and a 270 minute
+	// spendableDelay to be robust to periods of slow blocks.
+	return 3 * pendingBuffer(net)
+}
+
+// minBondLifetime gives the minimum bond lifetime (a duration from now until
+// lockTime) that a new bond should use to prevent a tier drop. bondExpiry is in
+// seconds.
+func minBondLifetime(net dex.Network, bondExpiry int64) time.Duration {
+	lifeTimeSec := 2*(pendingBuffer(net)+bondExpiry) + spendableDelay(net) // 2*(p+E)+m
+	return time.Second * time.Duration(lifeTimeSec)
+}
+
+func (c *Core) rotateBonds(ctx context.Context) {
 	// 1. Refund bonds with passed lockTime.
 	// 2. Move bonds that are expired according to DEX bond expiry into
 	//    expiredBonds (lockTime<lockTimeThresh).
-	// 3. TODO: Add bonds to keep N bonds active, according to some optional config.
+	// 3. Add bonds to keep N bonds active, according to target tier and max
+	//    bonded amount, posting before expiry of the bond being replaced.
 	now := time.Now().Unix()
 
 	type bondID struct {
 		assetID uint32
 		coinID  []byte
 	}
 
+	if !c.bondKeysReady() { // not logged in, and nextBondKey requires login to decrypt bond xpriv
+		return // nothing to do until wallets are connected on login
+	}
+
 	for _, dc := range c.dexConnections() {
-		lockTimeThresh := now // in case dex is down, expire (to refund) when lock time is passed
-		dc.cfgMtx.RLock()
-		if dc.cfg != nil {
-			lockTimeThresh += int64(dc.cfg.BondExpiry)
+		initialized, unlocked, _ := dc.acct.status()
+		if !initialized {
+			continue // view-only or temporary connection
+		}
+		// Account unlocked is generally implied by bondKeysReady, but we will
+		// check per-account before post since accounts can be individually
+		// locked. However, we must refund bonds regardless.
+
+		lockTimeThresh := now // in case dex is down, expire (to refund when lock time is passed)
+		var bondExpiry int64
+		bondAssets := make(map[uint32]*msgjson.BondAsset)
+		if cfg := dc.config(); cfg != nil {
+			bondExpiry = int64(cfg.BondExpiry)
+			for symb, ba := range cfg.BondAssets {
+				id, _ := dex.BipSymbolID(symb)
+				bondAssets[id] = ba
+			}
+			lockTimeThresh += int64(cfg.BondExpiry)
 		}
-		dc.cfgMtx.RUnlock()
+		replaceThresh := lockTimeThresh + pendingBuffer(c.net) // replace before expiry to avoid tier drop
```

Comment:
Nah, we just have to be at this spot to process refunds.  I'll double check that all the logic below excludes never connected dex (cfg == nil)

Reply:
`lockTimeThresh` and `replaceThresh` would be too early, which I guess means bonds **wouldn't** be expired or replaced when maybe they should be, so limited damage, but I'm not feeling great about running `filterExpiredBonds` with bad numbers.

Reply:
> which I guess means bonds **wouldn't** be expired or replaced when maybe they should be, so limited damage

Right, this is totally intended though.  In the absence of expiry info, don't move bonds into the expired bucket until we actually reach locktime. A few lines up this is doc'd:

``` go
lockTimeThresh := now // in case dex is down, expire (to refund when lock time is passed)
```

`replaceThresh` only affects creation of the `weak` slice, which is purely local to this method, no fields of `dc.acct` modified in doing that.  And if all important guards on the posting branch block any action:

```go
if mustPost > 0 && targetTier > 0 && bondExpiry > 0 {
	c.log.Infof("Gotta post %d bond increments now. Target tier %d, current tier %d (%d weak, %d pending)",
		mustPost, targetTier, tier, weakStrength, pendingStrength)
	if !unlocked || dc.status() != comms.Connected {
		c.log.Warnf("Unable to post the required bond while disconnected or logged out.")
		continue
	}
```

Here the absence of a `cfg` is implied by both of `bondExpiry` and `unlocked`, with `dc.status` being strongly correlated.

Happy to introduce a new flag to be extra safe though.